src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1041,102 +1041,184 @@ private bool IsValidFixedVariableInitializer(TypeSymbol declType, SourceLocalSym
+                    if (initializerType.IsArray())
                     {
-                        Error(diagnostics, ErrorCode.ERR_ManagedAddr, initializerSyntax, elementType);
-                        hasErrors = true;
+                        // See ExpressionBinder::BindPtrToArray (though most of that functionality is now in LocalRewriter).
+                        var arrayType = (ArrayTypeSymbol)initializerType;
+                        elementType = arrayType.ElementType;
+                        break;
                     }
-                }
-                else if (!initializerOpt.HasAnyErrors)
-                {
-                    elementType = initializerOpt.Type;
-                    switch (initializerOpt.Kind)
+
+                    // check for a special ref-returning method
+                    var additionalDiagnostics = DiagnosticBag.GetInstance();
+                    try
                     {
-                        case BoundKind.AddressOfOperator:
-                            elementType = ((BoundAddressOfOperator)initializerOpt).Operand.Type;
+                        getPinnableMethod = GetDangerousGetPinnableReferenceMethod(initializerOpt, additionalDiagnostics);
+
+                        bool extensisbleFixedEnabled = ((CSharpParseOptions)initializerOpt.SyntaxTree.Options)?.IsFeatureEnabled(MessageID.IDS_FeatureExtensibleFixedStatement) != false;
+
+                        // check for String
+                        // NOTE: We will allow DangerousGetPinnableReferenceMethod to take precendence, but only if it is a member of System.String
+                        if (initializerType.SpecialType == SpecialType.System_String &&
+                            ((object)getPinnableMethod == null || !extensisbleFixedEnabled || getPinnableMethod.ContainingType.SpecialType != SpecialType.System_String))
+                        {
+                            elementType = this.GetSpecialType(SpecialType.System_Char, diagnostics, initializerSyntax);
+                            Debug.Assert(!elementType.IsManagedType);
                             break;
-                        case BoundKind.FieldAccess:
-                            var fa = (BoundFieldAccess)initializerOpt;
-                            if (!fa.FieldSymbol.IsFixed)
+                        }
+
+                        if (extensisbleFixedEnabled)
+                        {
+                            // not a specially known type, check for getPinnableMethod
+                            if (getPinnableMethod != null)
                             {
-                                Error(diagnostics, ErrorCode.ERR_FixedNotNeeded, initializerSyntax);
-                                return true;
+                                elementType = getPinnableMethod.ReturnType;
+                                break;
                             }
-                            else
+                            else if (additionalDiagnostics != null)
```

Comment:
Seems like `additionalDiagnostics` is always set.

Reply:
not always. it is reported only on cases when we found the member, but something was wrong with it. 
The idea is to help implementers of the method, but not complain about the method when it is clearly not provided.

---
In reply to: [167649336](https://github.com/dotnet/roslyn/pull/24469#discussion_r167649336) [](ancestors = 167649336)